tests/GeneratedHydratorTest/CodeGenerator/Visitor/HydratorMethodsVisitorTest.php

```diff
@@ -81,4 +83,106 @@ public function classAstProvider() : array
             [$staticClassName, $parser->parse('<?php ' . $staticClassCode)[0], ['taz']],
         ];
     }
+
+	/**
+	 * @dataProvider propertyAstProvider
+	 */
+	public function testPropertyCodeGeneration(string $className, Class_ $classNode, array $properties) : void
```

Comment:
Hmm, this code is indeed complex and fragile. I applaud your efforts!

Still, I was wondering if we could write something like this:

```php
$expected = <<<'PHP'
public function hydrate(...)
{
   // ...
}
PHP;

$actual = <<<'PHP'
public function hydrate(...)
{
   // ...
}
PHP;

$expectedAst = $parser->parse($expected);
$actualAst = $parser->parse($actual);

self::assertEquals(
    $expected->some->path->to->the->method->node,
    $actual->some->path->to->the->method->node
);
```

Is such a test feasible, in your opinion?

Also feasible: using the PHP-Parser pretty-printer to compare the AST via string rather than AST.

Reply:
Was also thinking about using the pretty-printer and comparing strings :) though you would not like it ;)

Not sure whether assertEquals really works on a deeply nested array with different objects.. does it?